### PR TITLE
nixos/yazi: init

### PR DESCRIPTION
**Pull request by linsui**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

@XYenon @teto

### CODE REVIEW

File: nixos/modules/programs/yazi.nix [hunk 1]
```diff
@@ -0,0 +1,49 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
```
Comment (nagy) [inline nixos/modules/programs/yazi.nix:3]: Can you try to remove this? There is an effort to remove such wide-scoped `with` statements.

Comment (linsui) [reply nixos/modules/programs/yazi.nix]: Fixed.

File: nixos/modules/programs/yazi.nix [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      description = lib.mdDoc ''
+        Configuration included in `$YAZI_CONFIG_HOME`.
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    environment.systemPackages = [ cfg.package ];
+
+    environment.variables.YAZI_CONFIG_HOME = pkgs.linkFarm "yazi" (lib.attrsets.mergeAttrsList (map
```
Comment (misuzu) [inline nixos/modules/programs/yazi.nix:40]: > These variables will be set on shell initialisation (e.g. in /etc/profile). 

Each change to `settings` is going to need a relogin/shell restart. Maybe point it to a symlink in `/etc` instead?

File: nixos/modules/programs/yazi.nix [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      (name: lib.optionalAttrs (cfg.settings."${name}" != { }) {
```
Comment (jian-lin) [inline nixos/modules/programs/yazi.nix:41]:
```suggestion
      (name: lib.optionalAttrs (cfg.settings.${name} != { }) {
```

File: nixos/modules/programs/yazi.nix [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        "${name}.toml" = settingsFormat.generate "${name}.toml" cfg.settings."${name}";
```
Comment (jian-lin) [inline nixos/modules/programs/yazi.nix:42]:
```suggestion
        "${name}.toml" = settingsFormat.generate "${name}.toml" cfg.settings.${name};
```

**Review by jian-lin — APPROVED** [review]

File: nixos/modules/programs/yazi.nix (hunk 2 above)
Comment (linsui) [reply nixos/modules/programs/yazi.nix]: Good point!

File: nixos/modules/programs/yazi.nix (hunk 3 above)
Comment (linsui) [reply nixos/modules/programs/yazi.nix]: Fixed.

**Review by misuzu — APPROVED** [review]

**Comment by jian-lin** [comment]
According to the home-manager yazi module, there are also [shell integrations](https://github.com/nix-community/home-manager/blob/886ea1d213efd1082f419d066e89ef37635dc970/modules/programs/yazi.nix#L32).

Can that be added here?

**Comment by linsui** [comment]
I thought that's not provided by upstream. https://github.com/nix-community/home-manager/blob/886ea1d213efd1082f419d066e89ef37635dc970/modules/programs/yazi.nix#L9

**Comment by XYenon** [comment]
> I thought that's not provided by upstream. https://github.com/nix-community/home-manager/blob/886ea1d213efd1082f419d066e89ef37635dc970/modules/programs/yazi.nix#L9

It's from yazi's [README](https://github.com/sxyazi/yazi#usage)

**Comment by linsui** [comment]
Right. But it's just an example.

**Pull request merged**